app/javascript/mastodon/reducers/contexts.ts

```diff
@@ -0,0 +1,156 @@
+import type { Reducer } from '@reduxjs/toolkit';
+import type { List as ImmutableList } from 'immutable';
+
+import { timelineDelete } from 'mastodon/actions/timelines_typed';
+import type { ApiRelationshipJSON } from 'mastodon/api_types/relationships';
+import type {
+  ApiStatusJSON,
+  ApiContextJSON,
+} from 'mastodon/api_types/statuses';
+import type { Status } from 'mastodon/models/status';
+
+import { blockAccountSuccess, muteAccountSuccess } from '../actions/accounts';
+import { fetchContext } from '../actions/statuses';
+import { TIMELINE_UPDATE } from '../actions/timelines';
+import { compareId } from '../compare_id';
+
+interface State {
+  inReplyTos: Record<string, string | undefined>;
+  replies: Record<string, string[] | undefined>;
+}
+
+const initialState: State = {
+  inReplyTos: {},
+  replies: {},
+};
+
+const normalizeContext = (
+  state: State,
+  id: string,
+  { ancestors, descendants }: ApiContextJSON,
+) => {
+  const replies = { ...state.replies };
+  const inReplyTos = { ...state.inReplyTos };
```

Comment:
Haven't we moved to a model where we can simply mutate the state in-place? If so, that may be preferable to re-defining things like this.